drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c

```diff
@@ -0,0 +1,1808 @@
+					  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_vid_cap(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set capture format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_fmt_meta_cap(struct file *file, void *priv,
+					   struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_meta_cap(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set meta format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_fmt_vid_out(struct file *file, void *priv,
+					  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_vid_out(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set output format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_selection(struct file *file, void *fh,
+					struct v4l2_selection *s)
+{
+	struct mmal_parameter_crop_t crop;
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_dev *dev = node_get_dev(node);
+	struct vchiq_mmal_port *port = get_port_data(node);
+
+	/* This return value is required fro V4L2 compliance. */
+	if (NODE_IS_STATS(node))
+		return -ENOTTY;
+
+	if (!NODE_IS_OUTPUT(node) || s->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
+		return -EINVAL;
+
+	if (!s->r.width || !s->r.height)
+		return -EINVAL;
+
+	crop.rect.x = s->r.left;
+	crop.rect.y = s->r.top;
+	crop.rect.width = s->r.width;
+	crop.rect.height = s->r.height;
+
+	return vchiq_mmal_port_parameter_set(dev->mmal_instance, port,
+					     MMAL_PARAMETER_CROP,
+					     &crop, sizeof(crop));
+}
+
+static int bcm2835_isp_node_streamon(struct file *file, void *priv,
+				     enum v4l2_buf_type type)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "Stream on for node %s[%d]\n", node->name, node->id);
+
+	return vb2_streamon(&node->queue, type);
+}
+
+static int bcm2835_isp_node_streamoff(struct file *file, void *priv,
+				      enum v4l2_buf_type type)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "Stream off for node %s[%d]\n", node->name, node->id);
+
+	vb2_streamoff(&node->queue, type);
+	return 0;
+}
+
+static int bcm3285_isp_subscribe_event(struct v4l2_fh *fh,
+				       const struct v4l2_event_subscription *s)
+{
+	switch (s->type) {
+	/* Cannot change source parameters dynamically at runtime. */
+	case V4L2_EVENT_SOURCE_CHANGE:
+		return -EINVAL;
```

Comment:
Do we need to reject it then?
Let it pass into the default clause and the framework will handle it, but nothing will ever generate the event.

Reply:
It seems you do need to explicitly reject it.  Otherwise the compliance test fails if I let it pass through he default clause.

From what I can tell, the compliance test is expecting this specific behaviour for m2m devices.